src/devtools/client/webconsole/actions/messages.ts

```diff
@@ -231,3 +239,128 @@ export const messagesClearEvaluations = messageEvaluationsCleared;
 export const messagesClearLogpoint = logpointMessagesCleared;
 export const messageOpen = messageOpened;
 export const messageClose = messageClosed;
+
+function isNewFocusRegionSubset(
+  prevFocusRegion: FocusRegion | null,
+  nextFocusRegion: FocusRegion | null
+): boolean {
+  if (prevFocusRegion === null) {
+    // Previously the entire timeline was selected.
+    // No matter what the new focus region is, it will be a subset.
+    return true;
+  } else if (nextFocusRegion === null) {
+    // The new selection includes the entire timeline.
+    // No matter what the previous focus region is, the new one is not a subset.
+    return false;
+  } else {
+    return (
+      nextFocusRegion.startTime >= prevFocusRegion.startTime &&
+      nextFocusRegion.endTime <= prevFocusRegion.endTime
+    );
+  }
+}
+
+export function refetchMessages(focusRegion: FocusRegion | null): UIThunkAction {
+  return async (dispatch, getState, { ThreadFront }) => {
+    const state = getState();
+    const didOverflow = getConsoleOverflow(state);
+    const lastFetchedForFocusRegion = getLastFetchedForFocusRegion(state);
+
+    // Soft Focus: The frontend only needs to refetch data if:
+    // 1. The most recent time it requested data "overflowed" (too many messages to send them all), or
+    // 2. The new focus region is outside of the most recent region we fetched messages for.
+    //
+    // There are two things to note about the second bullet point above:
+    // 1. When devtools is first opened, there is no focused region.
+    //    This is equivalent to focusing on the entire timeline, so we often won't need to refetch messages when focusing for the first time.
+    // 2. We shouldn't compare the new focus region to the most recent focus region,
+    //    but rather to the most recent focus region that we fetched messages for (the entire timeline in many cases).
+    //    If we don't need to refetch after zooming in, then we won't need to refetch after zooming back out either,
+    //    (unless our fetches have overflowed at some point).
```

Comment:
:speech_balloon: ditto, love the explanation here - I think this will really help me a lot on my side

Reply:
This is all great. As mentioned in standup today and in https://github.com/RecordReplay/devtools/issues/6591, this logic is the same for all resources that could overflow, so we will likely want to reuse this particular check elsewhere at some point.

Reply:
Yeah, good call. Wasn't sure where to put this for now since it could apply to other resource types.

Maybe the answer is to move the subset helper method into `utils/timeline` so it can be referenced by other code that cares about such things, but keep the state (both overflow and `overflow` and `lastFetchedForFocusRegion`) in messages since they're very message-specific. (We might want to re-fetch some random resource even if we don't happen to need to refetch messages for example.)

Reply:
yep, that's exactly right. The logic for "given a new focus window, a most recently fetched window, and an overflow,  flag, should I fetch?" is the same across everything, but the "last fetched" and "overflow" portions are state that are unique to each resource that cares about this check.